### DISCUSSION TITLE
ESP8266 lwIP Implementation

### DISCUSSION THREAD

**Discussion by AmeyAvhad** [micropython/micropython#13188] [ESP8266]
Hey can anyone help me with how to set the **IP_TOS** value in **MQTT** messages in **ESP8266**. I need to change the IP_TOS value for my project  but the lwip stack doesn't seems to have the support for it, if so how can I implement it in the source code.

**Comment by jimmo** [comment]
@AmeyAvhad LWIP does support TOS, but the MicroPython bindings do not.

You need to add handling for it to setsockopt, see https://github.com/micropython/micropython/blob/master/extmod/modlwip.c#L1374, to call lwip's setsockopt.

**Reply by AmeyAvhad** [reply]
Hey thanks for the reply, I am fairly new to it so can you just tell me one thing that do I need to make changes to any other files excluding modlwip.c for it to work. And if possible, provide a sample code to handle it.

Thanks Once Again.

**Reply by jimmo** [reply]
I think it should be just `modlwip.c`

**Reply by AmeyAvhad** [reply]
Hey @jimmo , can you help its not working I added the code for IP_TOS function 

```c
case IP_TOS:{
mp_int_t val = mp_obj_get_int(args[3]);
            socket->tos = val;
            break;
}
``` 
and edited the lwip object
```c
typedef struct _lwip_socket_obj_t {
    mp_obj_base_t base;

    volatile union {
        struct tcp_pcb *tcp;
        struct udp_pcb *udp;
        struct raw_pcb *raw;
    } pcb;
    volatile union {
        struct pbuf *pbuf;
        struct {
            uint8_t alloc;
            uint8_t iget;
            uint8_t iput;
            union {
                struct tcp_pcb *item; // if alloc == 0
                struct tcp_pcb **array; // if alloc != 0
            } tcp;
        } connection;
    } incoming;
    mp_obj_t callback;
    byte peer[4];
    mp_uint_t peer_port;
    mp_uint_t timeout;
    uint16_t recv_offset;
    uint8_t tos;
    uint8_t domain;
    uint8_t type;

    #define STATE_NEW 0
    #define STATE_LISTENING 1
    #define STATE_CONNECTING 2
    #define STATE_CONNECTED 3
    #define STATE_PEER_CLOSED 4
    #define STATE_ACTIVE_UDP 5
    // Negative value is lwIP error
    int8_t state;
} lwip_socket_obj_t;
``` 
but after building the firmware it stills shows that in module 'IP_TOS' is not available.

**Reply by jimmo** [reply]
@AmeyAvhad You need to provide your own value of `IP_TOS` (like we do for e.g. `IP_ADD_MEMBERSHIP`).

Adding a `tos` field to the socket struct won't do anything. You need to set the `tos` field on the `tcp` (or `udp`) member of the pcb union.

**Reply by AmeyAvhad** [reply]
Hey @jimmo thanks for the suggestion I made some changes,
```c
case IP_TOS: {
	 mp_int_t tos_value = mp_obj_get_int(args[3]);
         socket->pcb.ip->tos = (uint8_t)tos_value;
			break;
	}
``` 

and the build seems fine but when checking with **wireshark** the **DSCP** value doesnt seems to change.
